### PR TITLE
Add more SetParam tests for Global

### PR DESCRIPTION
**Pull request by ami-GS**
## Description

More SetParam test set for global parameter #1215

## Testing

This is test

## Documentation

N/A

### CODE REVIEW

File: src/test/lib/ApiTest.cpp [hunk 1]
```diff
@@ -1813,10 +2113,151 @@ void QuicTestGlobalSetParam()
...
+                    QUIC_PARAM_GLOBAL_DATAPATH_PROCESSORS,
+                    sizeof(Data) + 1,
+                    &Data));
+        }
+
+        //
+        // one of data is bigger than the number of its' platform cpus
+        //
+        {
+            TestScopeLogger LogScope("one of data is bigger than the number of its' platform cpus");
```
Comment (nibanks) [inline src/test/lib/ApiTest.cpp:2145]:
```suggestion
        // one of data is bigger than the number of its platform cpus
        //
        {
            TestScopeLogger LogScope("one of data is bigger than the number of its platform cpus");
```

File: src/test/lib/ApiTest.cpp [hunk 2]
```diff
@@ -1813,10 +2113,151 @@ void QuicTestGlobalSetParam()
...
+            TEST_QUIC_SUCCEEDED(
+                MsQuic->SetParam(
+                    nullptr,
+                    QUIC_PARAM_GLOBAL_DATAPATH_PROCESSORS,
+                    sizeof(Data),
+                    &Data));
+        }
+    }
+
+#if QUIC_TEST_DATAPATH_HOOKS_ENABLED
```
Comment (nibanks) [inline src/test/lib/ApiTest.cpp:2185]: I don't think you should add tests for test hooks. They already get used.

File: src/test/lib/ApiTest.cpp [hunk 3]
```diff
@@ -1753,8 +1753,207 @@ class SecConfigTestContext {
...
+
 void QuicTestGlobalSetParam()
 {
+    //
+    // QUIC_PARAM_GLOBAL_RETRY_MEMORY_PERCENT
+    //
+    {
+        TestScopeLogger LogScope("QUIC_PARAM_GLOBAL_RETRY_MEMORY_PERCENT");
+        uint16_t Percent = 26;
+        TEST_QUIC_SUCCEEDED(
```
[the remark this answers is not in the document]
Comment (ami-GS) [reply src/test/lib/ApiTest.cpp:1949]: Is there any function to revert all global setting? or need to do manually?

[the remark this answers is not in the document]
Comment (nibanks) [reply src/test/lib/ApiTest.cpp:1949]: We don't really have a way to clear already set parameters, but I believe `ClearGlobalVersionListScope` does something similar to what you might want to do here. You just need to make a version that caches all settings, and then sets them back.

[the remark this answers is not in the document]
Comment (ami-GS) [reply src/test/lib/ApiTest.cpp:1949]: Thanks! I will check the function.

File: src/test/lib/ApiTest.cpp (hunk 2 above)
Comment (ami-GS) [reply src/test/lib/ApiTest.cpp]: Fixed, thanks!

File: src/test/lib/ApiTest.cpp [hunk 4]
```diff
@@ -1753,8 +1753,210 @@ class SecConfigTestContext {
     }
 };
 
+void SettingApplyTests(HQUIC Handle, uint32_t Param) {
+    {
+        QUIC_SETTINGS Settings{0};
+        Settings.MaxAckDelayMs = UINT32_MAX;
```
Comment (nibanks) [inline src/test/lib/ApiTest.cpp:1759]: Along with all these negative `UINT32_MAX` size tests, it's probably a good idea to add tests that verify that `QUIC_VAR_INT_MAX` does work.

File: src/test/lib/ApiTest.cpp [hunk 5]
```diff
@@ -1753,8 +1753,210 @@ class SecConfigTestContext {
...
+        MsQuicConnection Connection(Registration);
+        TEST_QUIC_SUCCEEDED(Connection.GetInitStatus());
+        TEST_QUIC_SUCCEEDED(
+            MsQuic->ConnectionStart(
+                Connection.Handle,
+                ClientConfiguration,
+                QUIC_ADDRESS_FAMILY_INET,
+                "localhost",
+                4433));
+        CxPlatSleep(1000); // bit slow to set MsQuicLib.InUse = TRUE
```
Comment (nibanks) [inline src/test/lib/ApiTest.cpp:1914]: I don't think this is required. Please verify and if it's not, please remove.

File: src/test/lib/ApiTest.cpp [hunk 6]
```diff
@@ -1753,8 +1753,210 @@ class SecConfigTestContext {
...
+    TestScopeLogger LogScope("QuicTestStatefulGlobalSetParam");
+    MsQuicRegistration Registration(true);
+    TEST_TRUE(Registration.IsValid());
+
+    //
+    // Set QUIC_PARAM_GLOBAL_LOAD_BALACING_MODE after connection start (MsQuicLib.InUse)
+    //
+    {
+        TestScopeLogger LogScope("Set QUIC_PARAM_GLOBAL_LOAD_BALACING_MODE after connection start (MsQuicLib.InUse)");
+        GlobalSettingScope ParamScope(QUIC_PARAM_GLOBAL_LOAD_BALACING_MODE);
```
Comment (nibanks) [inline src/test/lib/ApiTest.cpp:1900]: You should only need this helper if you plan to successfully set the parameter, right?

File: src/test/lib/TestHelpers.h [hunk 7]
```diff
@@ -112,6 +112,49 @@ struct ClearGlobalVersionListScope {
...
 
+//
+// Global parameter setting might affect other tests' behavior.
+// This sets back the original value
+//
+struct GlobalSettingScope {
+    uint32_t Parameter;
+    uint32_t BufferLength;
+    void* OriginalValue;
+    GlobalSettingScope(uint32_t Parameter) : Parameter(Parameter), BufferLength(0), OriginalValue(nullptr) {
```
Comment (nibanks) [inline src/test/lib/TestHelpers.h:123]: My preference would be to do it this way:
```suggestion
    uint32_t Parameter;
    uint32_t BufferLength {0};
    void* OriginalValue {nullptr};
    GlobalSettingScope(uint32_t Parameter) : Parameter(Parameter) {
```

File: src/test/lib/ApiTest.cpp (hunk 6 above)
Comment (ami-GS) [reply src/test/lib/ApiTest.cpp:1900]: Yes!

File: src/test/lib/ApiTest.cpp (hunk 5 above)
Comment (ami-GS) [reply src/test/lib/ApiTest.cpp:1914]: Got error without this sleep...

Comment (nibanks) [reply src/test/lib/ApiTest.cpp:1914]: Try changing it to `100` ms instead then.

Comment (ami-GS) [reply src/test/lib/ApiTest.cpp]: 100ms works. done!

**Review by nibanks — APPROVED** [review]

**Pull request merged**